### PR TITLE
perf recipes and Mcore DistOpt params

### PR DESCRIPTION
**Pull request by malay-nagda**
# What does this PR do ?

Add a one line overview of what this PR aims to accomplish.

**Collection**: [Note which collection this PR will affect]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to # (issue)

### CODE REVIEW

File: nemo/collections/llm/recipes/gpt3_175b.py [hunk 1]
```diff
@@ -0,0 +1,239 @@
...
+from megatron.core.distributed import DistributedDataParallelConfig
+from pytorch_lightning.callbacks.callback import Callback
+
+from nemo import lightning as nl
+from nemo.collections.llm.api import pretrain
+from nemo.collections.llm.gpt.data.mock import MockDataModule
+from nemo.collections.llm.gpt.model import GPTConfig175B, GPTModel
+from nemo.collections.llm.recipes.log.default import default_log, default_resume, tensorboard_logger
+from nemo.collections.llm.recipes.optim.adam import distributed_fused_adam_with_cosine_annealing
+from nemo.collections.llm.recipes.precision.mixed_precision import bf16_mixed, bf16_with_fp8_mixed
```
[the remark this answers is not in the document]
Comment (malay-nagda) [reply nemo/collections/llm/recipes/gpt3_175b.py:30]: @erhoo82 should the default recipe be bf16 or fp8?

File: nemo/collections/llm/gpt/model/base.py [hunk 2]
```diff
@@ -204,6 +207,9 @@ class GPTConfig5B(GPTConfig):
     ffn_hidden_size: int = 16384
     num_attention_heads: int = 32
 
+    bias_activation_fusion: bool = True
+    bias_dropout_add_fusion: bool = True
+
```
Comment (JimmyZhang12) [inline nemo/collections/llm/gpt/model/base.py:212]: Can these be removed since you already set these in base.py?

File: nemo/collections/llm/recipes/gpt3_175b.py [hunk 3]
```diff
@@ -0,0 +1,242 @@
...
+    recipe = pretrain_recipe(name=name, dir=dir, num_nodes=num_nodes, num_gpus_per_node=num_gpus_per_node, fn=fn)
+
+    recipe.trainer.callbacks.append(
+        run.Config(
+            MegatronCommOverlapCallback,
+            tp_comm_overlap=True,
+            tp_comm_overlap_cfg=userbuffers_fp8_h100_h12288_tp4_mbs1_seqlen2048,
+            defer_embedding_wgrad_compute=True,
+            wgrad_deferral_limit=50,
+            overlap_param_gather_with_optimizer_step=True,
```
Comment (JimmyZhang12) [inline nemo/collections/llm/recipes/gpt3_175b.py:237]: Can you add a comment about why these two are important for the user? and that they already get set automatically by the plugin

File: nemo/collections/llm/recipes/mixtral_8x22b.py [hunk 4]
```diff
@@ -217,7 +220,11 @@ def pretrain_recipe_performance(
     recipe = pretrain_recipe(name=name, dir=dir, num_nodes=num_nodes, num_gpus_per_node=num_gpus_per_node, fn=fn)
     recipe.trainer.callbacks.extend(
         [
-            run.Config(MegatronTokenDropCallback),
+            run.Config(
+                MegatronTokenDropCallback,
+                overlap_param_gather_with_optimizer_step=True,
```
Comment (JimmyZhang12) [inline nemo/collections/llm/recipes/mixtral_8x22b.py:225]: Thiese are attached to the wrong plugin?

File: nemo/collections/llm/gpt/model/base.py (hunk 2 above)
Comment (malay-nagda) [reply nemo/collections/llm/gpt/model/base.py:212]: This is in base.py. Do you mean some other file where this is set?

Comment (JimmyZhang12) [reply nemo/collections/llm/gpt/model/base.py:212]: Oh my mistake, could you set these in `GPTConfig`  instead?

Comment (malay-nagda) [reply nemo/collections/llm/gpt/model/base.py:212]: Ok. I'll set them here. But, also we have to set them to False in `NemotronConfig` then.

Comment (JimmyZhang12) [reply nemo/collections/llm/gpt/model/base.py]: Do they cause errors if you set them in Nemotron?

Comment (malay-nagda) [reply nemo/collections/llm/gpt/model/base.py]: Haven't tried. It is set to False for all nemotron models in Launcher.

File: nemo/collections/llm/recipes/gpt3_175b.py (hunk 3 above)
Comment (malay-nagda) [reply nemo/collections/llm/recipes/gpt3_175b.py:240]: Added

**Review by JimmyZhang12 — DISMISSED** [review]

File: nemo/collections/llm/recipes/gpt3_175b.py [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+        ),
+        data=run.Config(MockDataModule, seq_length=2048, global_batch_size=2048, micro_batch_size=2),
+        log=default_log(dir=dir, name=name, tensorboard_logger=tensorboard_logger(name=name)),
+        optim=distributed_fused_adam_with_cosine_annealing(max_lr=0.9e-4),
+        resume=default_resume(),
+    )
+
+
+@run.cli.factory(target=pretrain, name=NAME + "_performance")
+def pretrain_recipe_performance(
```
Comment (marcromeyn) [inline nemo/collections/llm/recipes/gpt3_175b.py:191]: I would be a bit afraid that this will blow up the number of recipes that we show in the CLI. Since we allow the user to call the recipe with arguments, what about adding a new arg to the pretrain recipes that will dictate if we use the performance one or not?

Comment (malay-nagda) [reply nemo/collections/llm/recipes/gpt3_175b.py:191]: I agree @marcromeyn . But, should we do this as a follow-up PR as this would involve performance recipes for all models (I assume) and the interface for perf recipe/perf mode would also change. Reviewing would be difficult/erroneous?

Also, cc: @erhoo82 @JimmyZhang12

Comment (erhoo82) [reply nemo/collections/llm/recipes/gpt3_175b.py:191]: @marcromeyn We need this performance recipe for every key model training because the baseline recipe is missing many performance features and it is hard to communicate with users without having a recipe that has all the important perf features. There is a high risk that a user can miss out many of them without asking us for the actual CLI command line every time.

Comment (erhoo82) [reply nemo/collections/llm/recipes/gpt3_175b.py:191]: @ericharper As we discussed, the baseline recipe is for debugging and we need a performance recipe for all model training.

BTW, are we supposed to have different recipes for different sequence lengths? 
Also, one question on the global batch size here.
image

Comment (ericharper) [reply nemo/collections/llm/recipes/gpt3_175b.py:191]: Sequence lengths have their own recipe, for example: https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/llm/recipes/llama3_70b_64k.py

Comment (erhoo82) [reply nemo/collections/llm/recipes/gpt3_175b.py:191]: @ericharper 
Yes, I see that it uses different a CP size for the target seqlen. So, I think it is only TP and CP mapping that needs to be changed. Do we need a full recipe only for this change? Also, I think the global batch size there is wrong.

Comment (ericharper) [reply nemo/collections/llm/recipes/gpt3_175b.py:191]: We decided previously to include the full recipe for this but we could revisit it. If there's a bug please send a fix.

@malay-nagda we can merge this as is, but please send a follow up PR with @marcromeyn 's suggestion

**Review by ericharper — DISMISSED** [review]
LGTM. Thanks!

File: nemo/collections/llm/gpt/model/base.py (hunk 2 above)
Comment (malay-nagda) [reply nemo/collections/llm/gpt/model/base.py:209]: This breaks other model tests in CI. Reverting to model specific fusion params.

**Review by JimmyZhang12 — APPROVED** [review]
